content/en/blog/_posts/2025-05-09-start-sidecar-first/index.md

```diff
@@ -0,0 +1,225 @@
+
+## A gentle refresher
+
+Let’s just remind that the [v1.29.0 release of Kubernetes](/blog/2023/12/13/kubernetes-v1-29-release/) added native support for [sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/), which can now be defined within the `spec.initContainers` field, but with `restartPolicy: Always` as in the example below (note that it is a partial snippet of the full Kubernetes manifest):
+
+```yaml
+initContainers:
+  - name: logshipper
+    image: alpine:latest
+    restartPolicy: Always
+    command: ['sh', '-c', 'tail -F /opt/logs.txt']
+    volumeMounts:
+    - name: data
+        mountPath: /opt
+```
+
+What’s the specifics of defining sidecars with an `spec.initContainers` block, rather than as a regular multi-container pod? `spec.initContainers` are always started first (before the main application) and terminated last (after the main application). Furthermore, when used with [Jobs](/docs/concepts/workloads/controllers/job/), sidecar would still be alive and potentially restarted after Job is done, and they will not block pod completion (see also [Pod sidecar containers tutorial](/docs/tutorials/configuration/pod-sidecar-containers/)).
+
+## The problem
+
+Now you know that defining a sidecar with this native approach will always start it before the main application. From the kubelet source code, we know that this often means [being started almost in parallel](https://medium.com/@marko.luksa/delaying-application-start-until-sidecar-is-ready-2ec2d21a7b74), and this is not always what we want to achieve. What we are really interested in is whether we can delay the start of the main application until the sidecar is not just started, but fully running and ready to serve.
+It might be a bit tricky because the problem with sidecars is there’s no obvious success signal, contrary to init containers - designed to run only for a specified period of time. With an init container, exit status 0 is unambiguously "I succeeded". With a sidecar, there are lots of points at which you can say "a thing is running".
+Starting one container only after the previous one is ready is part of a graceful deployment strategy, ensuring proper sequencing and stability during startup. It’s also actually how I’d expect sidecar containers to work as well, as it would cover the scenario where the main application is dependent on the sidecar. For example, it may happen that an app errors out if the sidecar isn’t available to serve requests (e.g., logging with DataDog). Sure, we could change the application code (and it would actually be the “best practice” solution), but sometimes we can’t - and this post focuses on this use case.
```

Comment:
> It’s also actually how I’d expect sidecar containers to work as well, as it would cover the scenario where the main application is dependent on the sidecar.

Consider rewording for better clarity.